File: lintcode/python/0691_recover_binary_search_tree.py
#!/usr/bin/python -t

"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""

class Solution:
    def __init__(self):
        self.first = None
        self.second = None
        self.pre = None
    """
    @param root: the given tree
    @return: the tree after swapping
    """
    def bstSwappedNode(self, root):
        # write your code here
        if not root:
            return root
            
        self.dfs(root)
        
        if self.first != None:
            self.first.val, self.second.val = self.second.val, self.first.val
            
        return root
        
    def dfs(self, node):
        if node == None:
            return
        
        self.dfs(node.left)
        
        if self.pre != None:
            if self.pre.val > node.val:
                if self.first == None:
                    self.first = self.pre
                self.second = node
            
        self.pre = node
        
        self.dfs(node.right)
        
        return
    

File: lintcode/golang/0691_recover_binary_search_tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

/**
 * @param root: the given tree
 * @return: the tree after swapping
 */
 
const MaxInt = int(^uint(0)>>1)
const MinInt = -MaxInt-1

var pre *TreeNode
var first, second *TreeNode

func bstSwappedNode (root *TreeNode) *TreeNode {
    // write your code here
    if root == nil {
        return nil
    }
    
    pre = &TreeNode{Val: MinInt}
    
    dfs(root)
    
    if first != nil  && second != nil {
        tmp := first.Val
        first.Val = second.Val
        second.Val = tmp
    }
    
    return root
}

func dfs(node *TreeNode) {
    if node == nil {
        return
    }
    
    dfs(node.Left)
    
    if pre.Val > node.Val {
        if first == nil {
            first = pre
        }
        second = node
    }
    
    pre = node
    
    dfs(node.Right)
    
    return
}
